### ISSUE TITLE
[Pose + Object detection] About multi class training and assessment

### ISSUE THREAD

**Issue by jjlee93** [ultralytics/ultralytics#16741]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello!
Thankfully, through the advice you guys gave me last time, I was able to train the model for the human class(pose) and other classes (only object detection).
The advice I received was to keep the person class unchanged in the .txt file, which contains class and coordinate(bbox + keypoint) information, and to fill all keypoint coordinates with zeros for the new class since it will only be used for object detection.

The issue I noticed when applying this method was that, although no errors occurred, as the epoch value increased, the mAP50(P) and mAP50*90(P) values did not improve. When training only on the person class, the mAP50(P) reached 0.87, but after adding the new class, it stagnated at 0.3 and did not increase. Could you provide any advice on this?

Also, is there a different approach or updated method I should consider instead of the training process I followed?

Thanks!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @jjlee93, thank you for reaching out and for your interest in Ultralytics 🚀!

If this is a 🐛 Bug Report, we kindly request a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us diagnose the issue more effectively.

If your inquiry relates to custom training ❓, please include additional details such as dataset examples and training logs. It’s essential to ensure you are adhering to our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Be part of the vibrant Ultralytics community where it suits you best! Join our [Discord](https://ultralytics.com/discord) 🎧 for real-time discussions, explore insights on our [Discourse](https://community.ultralytics.com), or engage with peers on our [Subreddit](https://reddit.com/r/ultralytics).

## Upgrade

Ensure you are using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be executed in any verified environment with preinstalled dependencies ([CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/)):

- **Notebooks** with GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

This CI badge should be green to indicate that all tests are passing. The continuous integration tests ensure operational correctness across all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on multiple platforms.

Please note, this is an automated response. An Ultralytics engineer will assist you shortly. If you have any urgent questions, feel free to continue the discussion here or on our community channels. 😊

**Comment by jjlee93** [comment]
Thank you for comment. right, but I think it will too much time to create a new model with two heads.
Is there any other options to check mAP50 properly within current method?

**Comment by Y-T-G** [comment]
You can try training a detection model first. And then train a keypoint estimation (pose) model using those weights, while freezing all the previous layers so that only keypoint head is trained.

```python
from ultralytics import YOLO

model = YOLO("yolo11n-pose.yaml").load("detect_model.pt")

results = model.train(..., freeze=22)
``` 

Make sure to change the `yaml` and `nc` value inside it to match your number of classes. Otherwise some weights wouldn't match.

**Comment by jjlee93** [comment]
The current method I am using involves a .yaml file with 3 classes (person and 2 other objects). The .txt file (label information) contains the following data. 
cl bbox                keypoint
0  x1,y1,x2,y2       x1,y1, ... x17,y17  <Human>
1  x1,y1,x2,y2       0,0, ... 0,0             <Obj 1>
2  x1,y1,x2,y2       0,0, ... 0,0             <Obj 2>
The reason I adopted this approach is that if keypoint information(coordinate values) for the classes are not provided(blank), an error occurs, so model training cannot proceed. 
In my opinion, even if I follow the method you suggested, the same issue will arise unless keypoint information is provided in the .txt file. Is there any trick to proceed with training without inputting keypoint coordinates for classes?

**Comment by glenn-jocher** [comment]
To handle classes without keypoints, you can fill the keypoint coordinates with zeros, as you've done. This approach is necessary to maintain the format. Unfortunately, there's no alternative to bypass this requirement in the current setup.

**Comment by jjlee93** [comment]
Thank you guys, I solved this issue.
I want to check each instance result because I input 0 for other objects and it showed all instance result in the record. So I can't check human's keypoint mAP go up correctly. 
Today, I used val function(ex. model.val() ) and found each instance result,  it worked correctly.

**Comment by manu13008** [comment]
Hi guys,

Very interisting approach! I am trying to do the same thing.
Just a few questions there :

Here is my data.yaml file : 
```
names: 
- platform
- ball
- human


# Number of classes
nc: 3 

# Key points configuration
kpt_shape: [20, 3]  # [(x,y,visible)]
flip_idx: [18,1,2,6,7,8,3,4,5,10,9,14,15,19,11,12,16,17,0,13]

```

I annotated my images using CVAT. Platform and ball are simple bbox. Human is keypoints object.
I would like to use yolo11n-pose model to perform both human pose and object detection tasks.
Here is a sample of my frames with data input for Yolo training step.
First line (0) is the platform coordinates
Second line (1) is the ball coordinates
Third line (2) is the human coordinates with keypoints (20 keypoints)

As far as I understood, I should add 20 "fake" 0 keypoints in each of the first 2 lines.

Should I change then the kpt_shape and flip_idx of my data.yaml ? 
Something like : 
kpt_shape: [60, 3]  # [(x,y,visible)]
flip_idx: [58,41,42,46,47,48,43,44,45,50,49,54,55,59,51,52,56,57,40,53]

to keep consistency between my configuration ?

```
0 0.490219 0.780701 0.200078 0.082403
1 0.521602 0.961326 0.015313 0.025403
2 0.501262 0.602236 0.135508 0.681583 0.473219 0.266597 2 0.485539 0.304847 2 0.493828 0.335292 2 0.453023 0.331319 2 0.469172 0.424861 2 0.502695 0.551250 2 0.535133 0.287389 2 0.530328 0.417514 2 0.521367 0.527319 2 0.473406 0.477403 2 0.529211 0.467917 0 0.449602 0.733208 2 0.433508 0.786125 2 0.546781 0.606389 2 0.550727 0.731667 2 0.569016 0.790889 2 0.513008 0.612708 2 0.502023 0.943028 2 0.500125 0.261444 2 0.459242 0.619528 2
```

Is there any pros and cons using this approach instead of another ? 

Thanks in advance :)

**Comment by Y-T-G** [comment]
> As far as I understood, I should add 20 "fake" 0 keypoints in each of the first 2 lines.

Yes

> Should I change then the kpt_shape and flip_idx of my data.yaml ?
Something like :
kpt_shape: [60, 3] # [(x,y,visible)]

No. Each object still has 20 keypoints.

> Should I keep the human index as 0 ? (human : 0 , plateform : 1 , ball : 2)

Doesn't matter

**Comment by manu13008** [comment]
@Y-T-G Thanks a lot , training worked !

To fully understand, is there a risk using the yolo11n-pose model as main model for both tasks could lead to a lack of precision for object detections compared to the yolo11n model normally used only for object detection ?

I saw also another approach consisting at modifying the model heads to perform both tasks. What would you recommend the most between these 2 methods ?

Thanks a lot again

**Comment by bit-scientist** [comment]
hello, @jjlee93, would you mind sharing your data.yaml, one sample image and its label file so that I can understand the idea fully?
